### ISSUE TITLE
Change ArenaPath to ArenaPaths

### ISSUE THREAD

**Issue by afritz1** [afritz1/OpenTESArena#243]
Since Arena is a DOS game, it doesn't really have a common location on modern filesystems. It would be nice to automatically support paths like the player's Steam folder now that Arena is on that storefront.

Current behavior: the engine looks in the `ArenaPath` for ACD.EXE, and if not found, looks for A.EXE.
Desired behavior: the engine iterates all `ArenaPaths` looking for ACD.EXE, before iterating them all again looking for A.EXE.

ArenaPaths in `options-default.txt` would come with a string that gets expanded to the correct Steam directory. Paths should be semicolon-separated.

I don't know how to look up the user's Steam folder but there's probably a way to get that environment variable on Windows/Linux/macOS.

**Comment by makemeunsee** [comment]
Hi,

> I don't know how to look up the user's Steam folder but there's probably a way to get that environment variable on Windows/Linux/macOS.

I've been looking this up, I'm not convinced there's a simple, reliable way to do it, beyond the easy cases. Steam itself can be installed anywhere, and it can manage multiple data folders. With the Steam install path, Steam config could be read and from it, the data folders extracted, but then that's code which would break whenever Steam changes anything in this regard.

Maybe having a good set of default paths, across all platforms, would work for 80+ % of the users, and then if the engine doesn't find Arena data going through all the default folders, it can point the user to the relevant config entry for manual edition.

**Comment by afritz1** [comment]
A not-quite-JSON format and registry keys don't sound appealing on the surface. I mean we can try but it's not in the spirit of this project to depend on anything more complex than like an environment variable.

If we can cover 90% of Windows players with a Steam install by providing
```
ArenaPaths=C:/Program Files (x86)/Steam/steamapps/common/The Elder Scrolls Arena/ARENA;D:/SteamLibrary/steamapps/common/The Elder Scrolls Arena/ARENA
```
and something similar for Linux and macOS, that seems reasonable to me.

Beyond that I would probably just want a file dialog to let the player point to their custom ARENA/ARENACD folder. Not sure how to implement that, I think it's either Qt or GTK, the former which is a big dependency I don't want, and the second isn't quite Windows native. Doing an in-engine dialog could work but I'd need a couple things like an open-source font and a way to render it with SDL. I wonder how Daggerfall Unity does it.

**Comment by afritz1** [comment]
I tried looking in the registry but it doesn't have my D drive SteamLibrary path, just the place where Steam the executable is installed on my C drive.

**Comment by afritz1** [comment]
Arena's app ID is 1812290 and its appmanifest_1812290.acf starts like this:
```
"AppState"
{
	"appid"		"1812290"
	"Universe"		"1"
	"LauncherPath"		"C:\\Program Files (x86)\\Steam\\steam.exe"
	"name"		"The Elder Scrolls: Arena"
	"StateFlags"		"4"
	"installdir"		"The Elder Scrolls Arena"
```
OpenTESArena would use that `installdir` to find `ARENA`.

**Comment by afritz1** [comment]
Yeah, I'm setting up some stuff on a branch, will advise in a while.

**Comment by afritz1** [comment]
Okay I made a `steam-path` branch. Checkout your own copy of that. Just worry about getting the path working on your preferred OS then I'll do a pass for the other OSes afterwards.

Looks like .vdf and .acf can be parsed the same so I've made a SteamKeyValueFile for that. It should check if the file is valid to parse; maybe it could count the number of open and close curly braces.

Feel free to use any `std` types and anything in C++17.

SteamUtils.h should eventually have some `std::vector<std::string> getLibraryDirectories()` so we can iterate all the places the `The Elder Scrolls Arena` installpath might be.

Later I think the ArenaPath behavior I want is, we use ArenaPath if it points to a valid ARENA/ARENACD directory, otherwise we try looking for the Steam install. Because we want Steam to be what most people use UNLESS they specify their own ArenaPath. Maybe don't need to change it to ArenaPaths yet.

**Comment by LukasThyWalls** [comment]
Hello.

Sometime ago I did an install setup for a game translation and just for fun I went a little bit over the top detecting automatically the paths, so I took some information about this.

For Steam, the more proper way to go is the commented, locating the _libraryfolders.vdf_ file in the Steam installation and cycle between all the Path info of each Steam library folder (as you can have more than one) and check if there is ACD.EXE at one of them. Summarising the paths:

> 32bit: HKEY_LOCAL_MACHINE\SOFTWARE\Valve\Steam
> 64bit: HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Valve\Steam
> 	--> InstallPath
> 
> %InstallPath%\steamapps\libraryfolders.vdf
> 
> For %each_libraryfolder%
> 	check %each_libraryfolder%\steamapps\common\The Elder Scrolls Arena\ARENA\ACD.EXE

Is the better way because not only can be used in Windows, also can be adapted to Linux or Mac (although Arena only install directly in Windows, unless you use Proton/Wine or something like that and I didn't research about that so I don't know how to do it there).

I planned to do it like that in my translation installation setup, but in the end I only try to check the main Steam installation library folder, because later I found a easy way to do it, although in the future could began to not to work: Steam is still creating Windows uninstall information for each Steam game so you can find there the path directly:

> HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Steam App 1812290
> --> InstallLocation
>
> (1812290 is the Arena Steam game ID)
>
> check %InstallLocation%\ARENA\ACD.EXE

Also my setup checked the GOG paths, and in this case is much easier because GOG puts the patch for each game installed in the Windows registry:

> 32bit: HKEY_LOCAL_MACHINE\SOFTWARE\GOG.com\Games\1435828982
> 64bit: HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\GOG.com\Games\1435828982
> --> "Path"
>
> (1435828982 is the Arena GOG game ID)
>
> check %Path%\ACD.EXE

Maybe It's useful for this topic.
